patch.go

```diff
@@ -67,14 +71,23 @@ func (c *C) setenv(name, val string, valOK bool) {
 
 // Mkdir makes a temporary directory and returns its name.
 //
-// The directory and its contents will be removed when
-// c.Done is called.
+// In Go >= 1.14 the directory and its contents will be removed when the test
+// and all its subtests complete.
+// In previous Go versions the directory will be removed when c.Done is called.
 func (c *C) Mkdir() string {
 	name, err := ioutil.TempDir("", "quicktest-")
 	c.Assert(err, Equals, nil)
-	c.Defer(func() {
+	cleanup := getCleanup(c)
+	cleanup(func() {
 		err := os.RemoveAll(name)
 		c.Check(err, Equals, nil)
 	})
 	return name
 }
+
+func getCleanup(c *C) func(func()) {
```

Comment:
Alternatively:

```
// cleanup uses Cleanup when it can, falling back to using Defer.
func (c *C) cleanup(f func()) {
     if tb, ok := c.TB.(cleaner); ok {
         tb.Cleanup(f)
    } else {
         c.Defer(f)
    }
}
```

which means you can just call `c.cleanup` directly instead of using `getCleanup` and then calling that.
Arguably clearer, and marginally more efficient too.

Reply:
Good idea, done.